core/src/client/importer.rs

```diff
@@ -134,40 +134,24 @@ impl Importer {
                 if !is_empty {
                     ctrace!(CLIENT, "Call new_blocks even though block verification queue is not empty");
                 }
-                let (enacted, retracted) = self.calculate_enacted_retracted(&import_results);
-                self.miner.chain_new_blocks(client, &imported_blocks, &invalid_blocks, &enacted, &retracted);
-                client.new_blocks(&imported_blocks, &invalid_blocks, &enacted, &retracted, &[]);
+                let enacted = self.extract_enacted(&import_results);
+                self.miner.chain_new_blocks(client, &imported_blocks, &invalid_blocks, &enacted);
+                client.new_blocks(&imported_blocks, &invalid_blocks, &enacted, &[]);
             }
         }
 
         client.db().flush().expect("DB flush failed.");
         imported
     }
 
-    pub fn calculate_enacted_retracted(&self, import_results: &[ImportRoute]) -> (Vec<BlockHash>, Vec<BlockHash>) {
-        fn map_to_vec(map: Vec<(BlockHash, bool)>) -> Vec<BlockHash> {
-            map.into_iter().map(|(k, _v)| k).collect()
-        }
-
-        // In ImportRoute we get all the blocks that have been enacted and retracted by single insert.
-        // Because we are doing multiple inserts some of the blocks that were enacted in import `k`
-        // could be retracted in import `k+1`. This is why to understand if after all inserts
-        // the block is enacted or retracted we iterate over all routes and at the end final state
-        // will be in the hashmap
-        let map = import_results.iter().fold(HashMap::new(), |mut map, route| {
-            for hash in &route.enacted {
-                map.insert(*hash, true);
-            }
-            for hash in &route.retracted {
-                map.insert(*hash, false);
-            }
-            map
+    pub fn extract_enacted(&self, import_results: &[ImportRoute]) -> Vec<BlockHash> {
+        let set = import_results.iter().fold(HashSet::new(), |mut set, route| {
+            route.enacted.iter().for_each(|hash| {
+                set.insert(*hash);
+            });
+            set
         });
-
-        // Split to enacted retracted (using hashmap value)
-        let (enacted, retracted) = map.into_iter().partition(|&(_k, v)| v);
-        // And convert tuples to keys
-        (map_to_vec(enacted), map_to_vec(retracted))
+        set.into_iter().collect()
```

Comment:
You can use `FromIterator` for it.

Reply:
Is there any noticeable difference between `set.into_iter().collect()` and `Vec::from_iter(set.into_iter())`?

Reply:
You can call `Vec::from_iter(set)` because `HashSet` implements`IntoIterator`.

Reply:
I modified the code